transport/peer/hostport/peer.go

```diff
@@ -0,0 +1,73 @@
+package hostport
+
+import (
+	"go.uber.org/yarpc/transport"
+
+	"github.com/uber-go/atomic"
+)
+
+// NewPeerIdentifier creates a new hostport.PeerIdentifier
+func NewPeerIdentifier(hostport string) PeerIdentifier {
+	return PeerIdentifier(hostport)
+}
+
+// PeerIdentifier uniquely references a host:port combination using a common interface
+type PeerIdentifier string
+
+// Identifier generates a (should be) unique identifier for this PeerIdentifier (to use in maps, etc)
+func (p PeerIdentifier) Identifier() string {
+	return string(p)
+}
+
+// NewPeer creates a new hostport.Peer from a hostport.PeerIdentifier, transport.Agent, and transport.PeerSubscriber
+func NewPeer(pid PeerIdentifier, agent transport.PeerAgent, subscriber transport.PeerSubscriber) *Peer {
+	return &Peer{
+		PeerIdentifier:   pid,
+		agent:            agent,
+		subscriber:       subscriber,
+		connectionStatus: transport.PeerAvailable,
+	}
+}
+
+// Peer keeps a subscriber to send status updates to it, and the PeerAgent that created it
+type Peer struct {
+	PeerIdentifier
+
+	agent            transport.PeerAgent
+	subscriber       transport.PeerSubscriber
```

Comment:
I ended up moving the subscriber list to the Agent because the agent is the only object that controls the lifetime of the object, when a peer has zero subscribers it should be removed from the agent (at that point it's dead weight).  But the only way to have those references deleted safely and atomically is to have the agent own the subscribers.

The "subscriber" here is the Agent, which is a bit confusing, maybe I should make a combined interface?